Set permissions for timestamp directory before publishing update to avoid permission denied

tsaarni opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug

#### What this PR does / why we need it:

This change fixes a race condition that was caused by setting the file owner, group and mode non-atomically, after the updated files had been published.
    
Users who were running non-root containers, without GID 0 permissions were getting intermittent `permission denied` errors if accessing the files during update, when they had removed read permissions from other users by setting `defaultMode: 0440` or similar and relied on `fsGroup` permissions for reading. This impacts secret and configmap volumes, and service account tokens on projected volumes.

This bug makes the system susceptible to a race condition and causes hard-to-troubleshoot intermittent failure of the application since the files likely have mandatory data for the application to function normally, such as TLS credentials. Applications are not likely prepared to retry opening the files several times since permission issues are not usually intermittent.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #114461

#### Special notes for your reviewer:

Following repository contains manual test procedure to reproduce the fault https://github.com/tsaarni/kubernetes-secret-race.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed file permission issues that happened during update of Secret/ConfigMap/projected volume when fsGroup is used. The problem caused a race condition where application gets intermittent permission denied error when reading files that were just updated, before the correct permissions were applied.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
N/A

tsaarni commented:
Hi @humblec and @msau42, could you please take a look at this PR? Thank you!

tsaarni commented:
I've updated the PR to cover also related race condition with token on projected volume.

furkatgofurov7 commented:
/ok-to-test

humblec commented:
Please check the unit test failure.

tsaarni commented:
> Please check the unit test failure.

Done! 

Since the PR proposes setting the final mode bits already in the payload struct, I now have changed the test to check for `0640`. This is the actual permission that has always been applied to files on disk - so no change from that perspective.  

Since the test case does not check for actual file permissions on disk, but instead only tests the payload struct, it checked for `0600` previously even though it was later overwritten with `0640` by `SetVolumeOwnership()` (this call is not included in the unit test).

tsaarni commented:
/retest

furkatgofurov7 commented:
> Please check the unit test failure.

@humblec can you please take a look again? Thanks. 
/cc @jsafrane @xing-yang @thockin hey folks! Can you please also take a look at this patch when you have some time, thank you!

thockin commented on pkg/volume/projected/projected.go:
```diff
@@ -321,12 +321,13 @@ func (s *projectedVolumeMounter) collectData(mounterArgs volume.MounterArgs) (ma
-			// change from 0600 to 0640.
 			mode := *s.source.DefaultMode
-			if mounterArgs.FsUser != nil || mounterArgs.FsGroup != nil {
+			if mounterArgs.FsUser != nil {
```
This block is confusing - what if both are set?  Can we leve some comments, please

tsaarni replied:
I've clarified the block a little: FsGroup shall take precedence (so outcome still follows [Case 1 here](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/1205-bound-service-account-tokens#proposed-heuristics)).

thockin reviewed (commented):
It feels like the real problem is that we finish the atomic `Write()` without calling `SetVolumeOwnership()`.  Should that be done BEFORE actually publishing the contents?  Maybe pass some callback into Write() ?

tsaarni commented:
> It feels like the real problem is that we finish the atomic `Write()` without calling `SetVolumeOwnership()`. Should that be done BEFORE actually publishing the contents? Maybe pass some callback into Write() ?

I was wondering that too. My thinking is: 

`SetVolumeOwnership()` fixes up permissions in the whole volume by recursively changing every file. Since we know exactly which files we create and which permissions they need, it seemed unnecessary to first create them with wrong permissions and then go through them again to correct them. 

Maybe secondary aspect is that `SetVolumeOwnership()` works on volume level. We know that the updated files are in the new timestamp directory. It seems bit redundant to go through the whole volume, including the old versions of the files that were already corrected during previous update, and will be removed at the last step of this update. 

Maybe there could be a new version of `SetVolumeOwnership()` that only touches the timestamp directory, but then again: why not set the correct permissions from the beginning? Doing the latter seemed minimal and maybe also less riskier change to fix the bug.

kashifest commented:
> but then again: why not set the correct permissions from the beginning? Doing the latter seemed minimal and maybe also less riskier change to fix the bug.

+1 on this. @thockin wdyt?

thockin commented:
Yeah, `SetVolumeOwnership()` is a little coarse for what we would want.  My fear is that we do something "manual" here and miss something or maybe later something changes there and it works everywhere EXCEPT here.  Doubly so because this same problem exists in projected, secret, configmap, and downwardapi volumes.  You can't just fix it in one place.

What if we changed `SetVolumeOwnership` to take an optional "subpath" argument and then changed `AtomicWriter.Write()` to take a "set permissions" function.  Something like:

```diff
diff --git a/pkg/volume/projected/projected.go b/pkg/volume/projected/projected.go
index a48567b2a6a..ef4b56b1680 100644
--- a/pkg/volume/projected/projected.go
+++ b/pkg/volume/projected/projected.go
@@ -230,15 +230,12 @@ func (s *projectedVolumeMounter) SetUpAt(dir string, mounterArgs volume.MounterA
 		return err
 	}
 
-	err = writer.Write(data)
-	if err != nil {
-		klog.Errorf("Error writing payload to dir: %v", err)
-		return err
+	setPerms := func(subPath string) error {
+		return volume.SetVolumeOwnership(s, subPath, mounterArgs.FsGroup, nil /*fsGroupChangePolicy*/, volumeutil.FSGroupCompleteHook(s.plugin, nil))
 	}
-
-	err = volume.SetVolumeOwnership(s, mounterArgs.FsGroup, nil /*fsGroupChangePolicy*/, volumeutil.FSGroupCompleteHook(s.plugin, nil))
+	err = writer.Write(data, setPerms)
 	if err != nil {
-		klog.Errorf("Error applying volume ownership settings for group: %v", mounterArgs.FsGroup)
+		klog.Errorf("Error writing payload to dir: %v", err)
 		return err
 	}
 	setupSuccess = true
diff --git a/pkg/volume/util/atomic_writer.go b/pkg/volume/util/atomic_writer.go
index 94428f6ffc8..4ff0fd1ba40 100644
--- a/pkg/volume/util/atomic_writer.go
+++ b/pkg/volume/util/atomic_writer.go
@@ -125,7 +125,7 @@ const (
 //
 // 10.  Old paths are removed from the user-visible portion of the target directory
 // 11.  The previous timestamped directory is removed, if it exists
-func (w *AtomicWriter) Write(payload map[string]FileProjection) error {
+func (w *AtomicWriter) Write(payload map[string]FileProjection, setPerms func(subPath string) error) error {
 	// (1)
 	cleanPayload, err := validatePayload(payload)
 	if err != nil {
@@ -182,6 +182,12 @@ func (w *AtomicWriter) Write(payload map[string]FileProjection) error {
 		klog.Errorf("%s: error writing payload to ts data directory %s: %v", w.logContext, tsDir, err)
 		return err
 	}
+	if setPerms != nil {
+		if err := setPerms(tsDirName); err != nil {
+			// FIXME: log an error
+			return err
+		}
+	}
 	klog.V(4).Infof("%s: performed write of new data to ts data directory: %s", w.logContext, tsDir)
 
 	// (7)
diff --git a/pkg/volume/volume_linux.go b/pkg/volume/volume_linux.go
index 57c02815029..cf463910520 100644
--- a/pkg/volume/volume_linux.go
+++ b/pkg/volume/volume_linux.go
@@ -40,7 +40,7 @@ const (
 // SetVolumeOwnership modifies the given volume to be owned by
 // fsGroup, and sets SetGid so that newly created files are owned by
 // fsGroup. If fsGroup is nil nothing is done.
-func SetVolumeOwnership(mounter Mounter, fsGroup *int64, fsGroupChangePolicy *v1.PodFSGroupChangePolicy, completeFunc func(types.CompleteFuncParam)) error {
+func SetVolumeOwnership(mounter Mounter, subPath string, fsGroup *int64, fsGroupChangePolicy *v1.PodFSGroupChangePolicy, completeFunc func(types.CompleteFuncParam)) error {
 	if fsGroup == nil {
 		return nil
 	}
@@ -55,7 +55,7 @@ func SetVolumeOwnership(mounter Mounter, fsGroup *int64, fsGroupChangePolicy *v1
 		return nil
 	}
 
-	err := walkDeep(mounter.GetPath(), func(path string, info os.FileInfo, err error) error {
+	err := walkDeep(filepath.Join(mounter.GetPath(), subPath), func(path string, info os.FileInfo, err error) error {
 		if err != nil {
 			return err
 		}
```

tsaarni committed:
Set permissions on volume before publishing update

This change fixes a race condition that was caused by setting the file owner,
group and mode non-atomically, after the updated files had been published.

Users who were running non-root containers, without GID 0 permissions, and
had removed read permissions from other users by setting defaultMode: 0440 or
similar, were getting intermittent permission denied errors when accessing
files on secret or configmap volumes or service account tokens on projected
volumes during update.

tsaarni commented:
@thockin I pushed a new commit according to your suggestion. However, I have left `SetVolumeOwnership()` as it was, even though I first commented about the granularity. My reasoning is that we should still set the permissions of all files at least during initial write.  I've also added a new test case for `setPerms()` call.

furkatgofurov7 commented:
/test pull-kubernetes-unit

thockin commented on pkg/volume/util/atomic_writer.go:
```diff
@@ -103,8 +103,10 @@ const (
-//     8.  The new data directory symlink is renamed to the data directory; rename is atomic
-//     9.  Symlinks and directory for new user-visible files are created (if needed).
+//
+//  8. Set the permissions of the files.
```
This whole comment block is weird, but if we are sticking to style, I'd prefer this as:

```
6.  The payload is written to the new timestamped directory
    7. Permissions are set (if setPerms is not nil) on the new timestamped directory and files
    8. A symlink to ...
    9. The new data directory symlink ...
    10. Symlinks and directory for new ...
```

In particular, I think setPerms should be after write, BEFORE symlink

tsaarni replied:
Fixed. 

The was some formatting that confused `gofmt` and it did not parse the numbered list anymore as numbered list properly.

I changed the order of setPerms and creation of `..data_tmp` symlink.

thockin commented on pkg/volume/util/atomic_writer.go:
```diff
@@ -193,6 +195,14 @@ func (w *AtomicWriter) Write(payload map[string]FileProjection) error {
 	// (8)
+	if setPerms != nil {
+		if err := setPerms(); err != nil {
+			klog.Errorf("%s: error applying volume ownership settings: %v", w.logContext, err)
```
drop "volume" here - it's minor but this makes it a smidge more general purpose

tsaarni replied:
Fixed!

thockin commented on pkg/volume/util/atomic_writer.go:
```diff
@@ -193,6 +195,14 @@ func (w *AtomicWriter) Write(payload map[string]FileProjection) error {
 
 	// (8)
+	if setPerms != nil {
+		if err := setPerms(); err != nil {
```
As a reader I am confused - if we don't pass the name of the dir we just made, what is `setPerms()` supposed to do?  Maybe we should pass subPath to the function and then the volumes can choose to disregard it if they want and do the whole volume instead (and they can comment why).

Can you add a note to the function comment explaining the contract with setPerms (and that it is optional).

tsaarni replied:
I've added the parameter and comment to volume implementations. I called it `tsDir` instead of `subPath` because while updating the test case I realized the path inside `Write()` was full path and not just the subdir.

thockin commented on pkg/volume/util/atomic_writer_test.go:
```diff
@@ -984,3 +985,46 @@ func TestCreateUserVisibleFiles(t *testing.T) {
 	}
 }
+
+func TestSetPerms(t *testing.T) {
```
thanks for a good test!

thockin reviewed (commented):
Looking pretty good!  Can you please:

1) retitle the PR
2) update the 1st PR comment
3) Write a good commit comment (maybe most of the same as the PR comment)
4) Shorten the commit first-line (git has a length limit :)

We're not always rgorous about commit hygiene but we are trying.

tsaarni commented:
> Looking pretty good! Can you please:
> 
> 1. retitle the PR
> 2. update the 1st PR comment
> 3. Write a good commit comment (maybe most of the same as the PR comment)
> 4. Shorten the commit first-line (git has a length limit :)
> 
> We're not always rgorous about commit hygiene but we are trying.

@thockin Thank you again for the review, much appreciated! I've now fixed the above.

In case needed, I've also published my test application and manual test procedure which I've been using, [available here](https://github.com/tsaarni/kubernetes-secret-race).

thockin commented on pkg/volume/configmap/configmap.go:
```diff
@@ -249,17 +249,17 @@ func (b *configMapVolumeMounter) SetUpAt(dir string, mounterArgs volume.MounterA
 	}
 
-	err = writer.Write(payload)
+	setPerms := func(tsDir string) error {
```
if you are going to ignore the argument, name it `_` which tells Go it can safely optimize and tells readers this was on purpose

tsaarni replied:
Fixed

thockin commented on pkg/volume/util/atomic_writer.go:
```diff
@@ -86,25 +86,35 @@ const (
 // directory.  Input paths must not begin with '..'.
+// setPerms is an optional pointer to a function that caller can provide to set the
+// permissions of the newly created files before they are published. The function is
+// passed tsDir which is full path to timestamp subdir.
```
full path or relative to the writer's root?

tsaarni replied:
Absolute path is now replaced with the timestamp directory name.

thockin commented on pkg/volume/util/atomic_writer.go:
```diff
@@ -185,14 +196,22 @@ func (w *AtomicWriter) Write(payload map[string]FileProjection) error {
 
 	// (7)
+	if setPerms != nil {
+		if err := setPerms(tsDir); err != nil {
```
I think it's simpler to pass `tsDirName` which is the basename

tsaarni replied:
I've replaced the absolute path with this one, renamed the parameter to `subPath` and changed the comment accordingly.

thockin reviewed (commented):
just nits left

thockin reviewed (commented):
Thanks!

/lgtm
/approve

furkatgofurov7 commented:
@thockin hi, thanks a lot for the review. Can this PR be considered to be backported to older branches? 
cc @kubernetes/sig-release-leads

thockin commented:
I could make an argument for 1.26 but I don't know if going back much farther is worth the risk.  It's not high-urgency (it self-resolves and it has not come up as a problem before), so if I were branch manager I would probably decline.

furkatgofurov7 commented:
> I could make an argument for 1.26 but I don't know if going back much farther is worth the risk.

@thockin thank you, I guess backporting it to the latest release branch is safe and considering your comment, I have created a backport for 1.26 here: https://github.com/kubernetes/kubernetes/pull/114728 PTAL.
